src/utilities/caching/getMemoryInternals.ts

```diff
@@ -0,0 +1,181 @@
+      typeof _getInMemoryCacheMemoryInternals
+    >)
+  : undefined;
+
+/**
+ * For internal purposes only - please call `ApolloClient.getMemoryInternals` instead
+ * @internal
+ */
+export const getApolloCacheMemoryInternals =
+  __DEV__ ?
+    (_getApolloCacheMemoryInternals as RemoveThis<
+      typeof _getApolloCacheMemoryInternals
+    >)
+  : undefined;
+
+function _getApolloClientMemoryInternals(this: ApolloClient<any>) {
+  if (!__DEV__) throw new Error("only supported in development mode");
+
+  return {
+    limits: cacheSizes,
+    sizes: {
+      global: {
+        print: globalCaches.print?.(),
+        parser: globalCaches.parser?.(),
+        canonicalStringify: globalCaches.canonicalStringify?.(),
+      },
+      links: linkInfo(this.link),
+      queryManager: {
+        Transforms: this["queryManager"]["transformCache"].size,
+        documentTransforms: transformInfo(
+          this["queryManager"].documentTransform
+        ),
+      },
+      cache: this.cache.getMemoryInternals?.(),
+    },
+  };
+}
+
+function _getApolloCacheMemoryInternals(this: ApolloCache<any>) {
+  return {
+    fragmentQueryDocuments: getWrapperInformation(this["getFragmentDoc"]),
+  };
+}
+
+function _getInMemoryCacheMemoryInternals(this: InMemoryCache) {
+  const fragments = this.config.fragments as
+    | undefined
+    | {
+        findFragmentSpreads?: Function;
+        transform?: Function;
+        lookup?: Function;
+      };
+
+  return {
+    ..._getApolloCacheMemoryInternals.apply(this as any),
+    addTypenameTransform: transformInfo(this["addTypenameTransform"]),
+    storeReader: {
+      executeSelectionSet: getWrapperInformation(
+        this["storeReader"]["executeSelectionSet"]
+      ),
+      executeSubSelectedArray: getWrapperInformation(
+        this["storeReader"]["executeSubSelectedArray"]
+      ),
+    },
+    maybeBroadcastWatch: getWrapperInformation(this["maybeBroadcastWatch"]),
+    fragmentRegistry: {
+      findFragmentSpreads: getWrapperInformation(
+        fragments?.findFragmentSpreads
+      ),
+      lookup: getWrapperInformation(fragments?.lookup),
+      transform: getWrapperInformation(fragments?.transform),
+    },
+  };
+}
+
+function isWrapper(f?: Function): f is OptimisticWrapperFunction<any, any> {
+  return !!f && "dirtyKey" in f;
+}
+
+function getWrapperInformation(f?: Function) {
+  return isWrapper(f) ? f.size : undefined;
+}
+
+function transformInfo(transform?: DocumentTransform): number[] {
+  return !transform ?
+      []
+    : [
+        getWrapperInformation(transform?.["performWork"]),
+        ...transformInfo(transform?.["left"]),
+        ...transformInfo(transform?.["right"]),
+      ].filter((x): x is number => x != null);
+}
+
+function linkInfo(link?: ApolloLink): unknown[] {
+  return !link ?
+      []
+    : [
+        link?.getMemoryInternals?.(),
+        ...linkInfo(link?.left),
+        ...linkInfo(link?.right),
+      ].filter((x) => x != null);
```

Comment:
```suggestion
      ].filter(Boolean);
```

might also be a nice shortcut that saves a couple bytes 🙂.

Reply:
I had that, but it also removed `0` values in the `DocumentTransform` use case further down (which just returns a number, not an object). 
So we definitely need this down there. I can see if extracting it out can save a few bytes so at least we don't repeat the function definition.

Reply:
Results in the same size, but slightly nicer: https://github.com/apollographql/apollo-client/pull/11409/commits/e283de3efc4ef51e07047a26f96b1629024406ca